File: app/src/test/java/com/example/mike/mp3player/service/library/ContentManagerTest.java
package com.example.mike.mp3player.service.library;

import android.support.v4.media.MediaBrowserCompat.MediaItem;

import com.example.mike.mp3player.commons.MediaItemType;
import com.example.mike.mp3player.service.library.content.request.ContentRequest;
import com.example.mike.mp3player.service.library.content.request.ContentRequestParser;
import com.example.mike.mp3player.service.library.content.retriever.ContentRetriever;
import com.example.mike.mp3player.service.library.content.retriever.RootRetriever;
import com.example.mike.mp3player.service.library.content.searcher.ContentSearcher;

import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import edu.emory.mathcs.backport.java.util.Collections;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ContentManagerTest {

    private static final String VALID_QUERY = "query";

    private ContentManager contentManager;

    @Mock
    private ContentRequestParser contentRequestParser;

    @Mock
    private RootRetriever rootRetriever;

    private static Map<MediaItemType, ContentSearcher> contentSearcherMap;

    @BeforeClass
    public static void setupClass() {
        MediaItem song1 = mock(MediaItem.class);
        MediaItem song2 = mock(MediaItem.class);
        List<MediaItem> songs = new ArrayList<>();
        ContentSearcher songSearcher = getContentSearch(MediaItemType.SONG, songs);
        songs.add(song1);
        songs.add(song2);
        MediaItem folder1 = mock(MediaItem.class);
        List<MediaItem> folders = new ArrayList<>();
        folders.add(folder1);
        ContentSearcher folderSearcher = getContentSearch(MediaItemType.FOLDER, folders);

        contentSearcherMap = new HashMap<>();
        contentSearcherMap.put(songSearcher.getSearchCategory(), songSearcher);
        contentSearcherMap.put(folderSearcher.getSearchCategory(), folderSearcher);
    }

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void testGetChildren() {
        final String id = "id";
        final List<MediaItem> expectedList = new ArrayList<>();
        final ContentRetriever contentRetriever = mock(ContentRetriever.class);
        ContentRequest contentRequest = new ContentRequest(null, id, null);
        when(contentRequestParser.parse(id)).thenReturn(contentRequest);
        when(contentRetriever.getChildren(contentRequest)).thenReturn(expectedList);

        Map<String, ContentRetriever> idToContentRetrieverMap = Collections.singletonMap(id, contentRetriever);
        this.contentManager = new ContentManager(idToContentRetrieverMap,
                null,
                contentRequestParser,
                rootRetriever);

        List<MediaItem> result =  contentManager.getChildren(id);
        assertEquals(expectedList, result);
    }

    @Test
    public void testGetChildrenNull() {
        final String incorrectId = "incorrectId";
        ContentRequest contentRequest = new ContentRequest(null, incorrectId, null);
        when(contentRequestParser.parse(incorrectId)).thenReturn(contentRequest);
        Map<String, ContentRetriever> idToContentRetrieverMap = new HashMap<>();
        this.contentManager = new ContentManager(idToContentRetrieverMap,
                null,
                contentRequestParser,
                rootRetriever);

        List<MediaItem> result =  contentManager.getChildren(incorrectId);
        assertNull(result);
    }

    /**
     * EXPECTED:
     * 1) SONGS_TITLE
     * 2) SONG 1
     * 3) SONG 2
     * 4) FOLDERS TITLE
     * 5) FOLDER 1
     * i.e result size 5
     */
    @Test
    public void testValidSearchQuery() {
        testSearch(VALID_QUERY, 5);
    }

    @Test
    public void ValidSearchWithWhiteSpace() {
        final String queryWithTrailingWhitespace = "   " + VALID_QUERY + "       ";
        testSearch(queryWithTrailingWhitespace, 5);
    }

    private void testSearch(String query, int expectedResultsSize) {
        this.contentManager = new ContentManager(null,
                contentSearcherMap,
                contentRequestParser,
                rootRetriever);

        List<MediaItem> result =  contentManager.search(query);
        assertNotNull(result);
        int resultSize = result.size();
        assertEquals(expectedResultsSize, resultSize);
    }

    private static ContentSearcher getContentSearch(MediaItemType type, List<MediaItem> result) {
        ContentSearcher contentSearcher = mock(ContentSearcher.class);
        when(contentSearcher.getSearchCategory()).thenReturn(type);
        when(contentSearcher.search(VALID_QUERY)).thenReturn(result);
        return contentSearcher;
    }

}

File: service/src/testFullDebug/java/com/github/goldy1992/mp3player/service/library/ContentManagerTest.kt
package com.github.goldy1992.mp3player.service.library

import androidx.media3.common.MediaItem
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.service.library.content.ContentRetrievers
import com.github.goldy1992.mp3player.service.library.content.request.ContentRequest
import com.github.goldy1992.mp3player.service.library.content.request.ContentRequestParser
import com.github.goldy1992.mp3player.service.library.content.retriever.ContentRetriever
import com.github.goldy1992.mp3player.service.library.content.retriever.MediaItemFromIdRetriever
import com.github.goldy1992.mp3player.service.library.content.retriever.RootRetriever
import com.github.goldy1992.mp3player.service.library.content.retriever.SongFromUriRetriever
import com.github.goldy1992.mp3player.service.library.content.searcher.ContentSearcher
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.*
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
@OptIn(ExperimentalCoroutinesApi::class)
class ContentManagerTest {
    private var contentManager: ContentManager? = null

    private val contentRequestParser: ContentRequestParser = mock<ContentRequestParser>()

    private val contentRetrievers: ContentRetrievers = mock<ContentRetrievers>()

    private val contentSearchers: ContentSearchers = mock<ContentSearchers>()
    
    private val rootRetriever: RootRetriever = mock<RootRetriever>()
    
    private val rootItem: MediaItem = mock<MediaItem>()
    
    private val mediaItemFromIdRetriever: MediaItemFromIdRetriever = mock<MediaItemFromIdRetriever>()
    
    private val songFromUriRetriever: SongFromUriRetriever = mock<SongFromUriRetriever>()

    private val testScheduler = TestCoroutineScheduler()
    private val dispatcher  = StandardTestDispatcher(testScheduler)
    private val testScope = TestScope(dispatcher)


    @Before
    fun setup() {

        whenever(contentRetrievers.root).thenReturn(rootRetriever)
        whenever(rootRetriever.getRootItem(any<MediaItemType>())).thenReturn(rootItem)
        contentManager = ContentManager(contentRetrievers,
                contentSearchers,
                contentRequestParser,
                songFromUriRetriever,
                mediaItemFromIdRetriever)
    }

    @Test
    fun testGetChildren() {
        val contentRetrieverId = "id"
        val expectedList: List<MediaItem> = ArrayList()
        val contentRetriever = mock<ContentRetriever>()
        whenever(contentRetrievers[contentRetrieverId]).thenReturn(contentRetriever)
        val contentRequest = ContentRequest("", contentRetrieverId, null)
        whenever(contentRequestParser.parse(contentRetrieverId)).thenReturn(contentRequest)
        whenever(contentRetriever.getChildren(contentRequest)).thenReturn(expectedList)
        val result = contentManager!!.getChildren(contentRetrieverId)
        Assert.assertEquals(expectedList, result)
    }

    @Test
    fun testGetChildrenNull() {
        val incorrectId = "incorrectId"
        val contentRequest = ContentRequest("", incorrectId, null)
        whenever(contentRequestParser.parse(incorrectId)).thenReturn(contentRequest)
        val result = contentManager!!.getChildren(incorrectId)
        Assert.assertNull(result)
    }

    /**
     * EXPECTED:
     * 1) SONGS_TITLE
     * 2) SONG 1
     * 3) SONG 2
     * 4) FOLDERS TITLE
     * 5) FOLDER 1
     * i.e result size 5
     */
    @Test
    fun testValidSearchQuery() = runTest(dispatcher) {
        testSearch(LOWER_CASE_VALID_QUERY, 5)
    }

    @Test
    fun validSearchWithWhiteSpace() = runTest(dispatcher) {
        val queryWithTrailingWhitespace = "   $VALID_QUERY       "
        testSearch(queryWithTrailingWhitespace, 5)
    }

    private suspend fun testSearch(query: String, expectedResultsSize: Int)  {
        val song1 = mock<MediaItem>()
        val song2 = mock<MediaItem>()
        val songs: MutableList<MediaItem> = ArrayList()
        songs.add(song1)
        songs.add(song2)
        val songSearcher = getContentSearch(MediaItemType.SONGS, songs)
        val folder1 = mock<MediaItem>()
        val folders: MutableList<MediaItem> = ArrayList()
        folders.add(folder1)
        val folderSearcher = getContentSearch(MediaItemType.FOLDER, folders)
        val contentSearcherList: MutableList<ContentSearcher> = ArrayList()
        contentSearcherList.add(songSearcher)
        contentSearcherList.add(folderSearcher)
        whenever(contentSearchers.all).thenReturn(contentSearcherList)
        whenever(contentSearchers[songSearcher.searchCategory]).thenReturn(songSearcher)
        whenever(contentSearchers[folderSearcher.searchCategory]).thenReturn(folderSearcher)
        val result = contentManager!!.search(query)
        Assert.assertNotNull(result)
        val resultSize = result.size
        Assert.assertEquals(expectedResultsSize.toLong(), resultSize.toLong())
    }

    private suspend fun getContentSearch(type: MediaItemType, result: List<MediaItem>): ContentSearcher {
        val contentSearcher = mock<ContentSearcher>()
        whenever<Any?>(contentSearcher.searchCategory).thenReturn(type)
        whenever(contentSearcher.search(VALID_QUERY)).thenReturn(result.toMutableList())
        return contentSearcher
    }

    companion object {
        private const val VALID_QUERY = "QUERY"
        private const val LOWER_CASE_VALID_QUERY = "query"
        private val contentSearcherMap: Map<MediaItemType, ContentSearcher>? = null
    }
}